### PR TITLE
Campaign crud 

### PR DESCRIPTION
**Pull request by ferhatelmas**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -2063,3 +2063,62 @@ export type User<UserType = UnknownType> = UserType & {
   teams?: string[];
   username?: string;
 };
+
+export type SegmentData = {
+  description: string;
+  filter: {
+    channel?: object;
+    user?: object;
```
Comment (mahboubii) [inline src/types.ts:2071]: are these enriched objects from API?

File: src/types.ts [hunk 2]
```diff
@@ -2063,3 +2063,62 @@ export type User<UserType = UnknownType> = UserType & {
...
+  id: string;
+  updated_at: string;
+  recipients?: number;
+} & SegmentData;
+
+export type CampaignData = {
+  attachments: Attachment[];
+  defaults: { [key: string]: string };
+  segment_id: string;
+  sender_id: string | null;
```
Comment (mahboubii) [inline src/types.ts:2088]: should this be optional to be omitted from the request?

File: src/types.ts [hunk 3]
```diff
@@ -2063,3 +2063,62 @@ export type User<UserType = UnknownType> = UserType & {
...
+  app_pk: number;
+  created_at: string;
+  id: string;
+  updated_at: string;
+  recipients?: number;
+} & SegmentData;
+
+export type CampaignData = {
+  attachments: Attachment[];
+  defaults: { [key: string]: string };
```
Comment (mahboubii) [inline src/types.ts:2086]:
```suggestion
  defaults: Record<string, string>;
```

**Review by mahboubii — COMMENTED** [review]
Should we also add some jsdoc for these new endpoints?

File: src/types.ts (hunk 1 above)
Comment (gumuz) [reply src/types.ts:2071]: no, they're either a User or Channel filter (like queryusers/querychannels), example:

`{"id": 1223, "username": {$eq: "amin"}}`

File: src/types.ts (hunk 2 above)
Comment (gumuz) [reply src/types.ts:2088]: Yeah, so I guess it should be `string?` instead. I sthat what you're suggesting?

File: src/types.ts (hunk 1 above)
Comment (mahboubii) [reply src/types.ts:2071]: in that case let's pull @nhannah to help us with this type

File: src/types.ts (hunk 2 above)
Comment (mahboubii) [reply src/types.ts]: exactly

**Review by mahboubii — APPROVED** [review]
🌮

**Review by gumuz — APPROVED** [review]

File: src/types.ts [hunk 4]
```diff
@@ -2063,3 +2063,63 @@ export type User<UserType = UnknownType> = UserType & {
...
+} & SegmentData;
+
+export type CampaignData = {
+  attachments: Attachment[];
+  defaults: Record<string, string>;
+  segment_id: string;
+  text: string;
+  push_notifications?: boolean;
+  sender_id?: string;
+  track_opened?: boolean;
```
Comment (ferhatelmas) [inline src/types.ts:2092]: we don't have this

File: src/types.ts [hunk 5]
```diff
@@ -2063,3 +2063,63 @@ export type User<UserType = UnknownType> = UserType & {
...
+    | 'draft'
+    | 'stopped'
+    | 'scheduled'
+    | 'completed'
+    | 'failed'
+    | 'canceled'
+    | 'in_progress';
+  completed_at?: string;
+  failed_at?: string;
+  progress?: {
```
Comment (ferhatelmas) [inline src/types.ts:2107]: this isn't aligned with api either, we might want to have this way, more user friendly

File: src/types.ts [hunk 6]
```diff
@@ -2063,3 +2063,63 @@ export type User<UserType = UnknownType> = UserType & {
...
+  stopped_at?: string;
+};
+
+export type Campaign = {
+  app_pk: string;
+  created_at: string;
+  id: string;
+  updated_at: string;
+  test?: boolean;
+  test_original_id?: string;
```
Comment (ferhatelmas) [inline src/types.ts:2123]: these flags aren't there either

**Pull request merged**